tests/templates/test_broadcast.py

```diff
@@ -288,3 +288,49 @@ def test_wire_sequence_generating_functions(self, function, wires, target):
 
         sequence = function(wires)
         assert sequence == target
+
+
+class TestCustomPattern:
+    """Additional tests for using broadcast with a custom pattern."""
+
+    @pytest.mark.parametrize("custom_pattern, pattern", [([[0, 1], [1, 2], [2, 3], [3, 0]], "ring"),
+                                                         ([[0, 1], [1, 2], [2, 3]], "chain"),
+                                                         ([[0, 1], [2, 3]], "double")
+                                                         ])
+    def test_reproduce_builtin_patterns(self, custom_pattern, pattern):
+        """Tests that the custom pattern can reproduce the built in patterns."""
+
+        dev = qml.device('default.qubit', wires=4)
+
+        # qnode using custom pattern
+        @qml.qnode(dev)
+        def circuit1():
+            broadcast(unitary=qml.CNOT, pattern=custom_pattern, wires=range(4))
+            return [qml.expval(qml.PauliZ(wires=w)) for w in range(4)]
+
+        # qnode using built-in pattern
+        @qml.qnode(dev)
+        def circuit2():
+            broadcast(unitary=qml.CNOT, pattern=pattern, wires=range(4))
+            return [qml.expval(qml.PauliZ(wires=w)) for w in range(4)]
+
+        custom = circuit1()
+        built_in = circuit2()
+        assert np.allclose(custom, built_in)
+
+    @pytest.mark.parametrize("custom_pattern", [1,
+                                                [1, 2],
+                                                [['a'], ['b']]
+                                                ])
+    def test_exception_custom_pattern_not_valid(self, custom_pattern):
+        """Tests that an exception is raised if the pattern is not a list of lists of integers."""
+
+        dev = qml.device('default.qubit', wires=2)
+
+        @qml.qnode(dev)
+        def circuit():
+            broadcast(unitary=qml.Hadamard, wires=[0, 1], pattern=custom_pattern)
+            return qml.expval(qml.PauliZ(0))
+
+        with pytest.raises(ValueError, match="a custom pattern must be a list"):
+            circuit()
```

Comment:
Nice tests! For a unit test, you could do something that tests that the gates were applied to the correct wires, by using the operation recorder:

```python
def circuit():
    broadcast(unitary=qml.CNOT, pattern=custom_pattern, wires=range(4))

with OperatonRecorder() as rec:
    circuit()

assert rec.queue[0].name == "CNOT"
assert rec.queue[0].wires = custom_pattern[0]
```
etc.

Reply:
Oh yes, I wanted to add that one!

Reply:
Decided to do one where I check the output instead!